### PR TITLE
Provide macros for error checking

### PR DESCRIPTION
**Pull request by brryan**
## PR Summary

Provide simple, informative macros `PARTHENON_REQUIRE(condition, message)` and `PARTHENON_FAIL(message)` as an alternative to throwing exceptions. These macros print some output that includes `__FILE__` and `__LINE__` macros, removing the need to include some identifying information in some error messages that could otherwise be ambiguous. 

Related to #81.

## PR Checklist

- [x] Code passes cpplint
- [x] New features are documented.
- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted

### CODE REVIEW

File: src/utils/error_checking.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+namespace parthenon {
+namespace ErrorChecking {
+
+inline void require(std::string const &condition, std::string const &message,
+                    std::string const &filename, int const linenumber) {
+  std::cout << "!!! PARTHENON ERROR !!!" << std::endl;
+  std::cout << "  Condition:   " << condition << std::endl;
+  std::cout << "  Message:     " << message << std::endl;
+  std::cout << "  File:        " << filename << std::endl;
+  std::cout << "  Line number: " << linenumber << std::endl;
```
Comment (Yurlungur) [inline src/utils/error_checking.hpp:41]: This should probably be `std::cerr` instead of `std::cout`.

File: src/utils/error_checking.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+inline void fail(std::string const &message, std::string const &filename,
+                 int const linenumber) {
+  std::cout << "!!! PARTHENON ERROR !!!" << std::endl;
+  std::cout << "  Message:     " << message << std::endl;
+  std::cout << "  File:        " << filename << std::endl;
+  std::cout << "  Line number: " << linenumber << std::endl;
```
Comment (Yurlungur) [inline src/utils/error_checking.hpp:49]: Here too.

**Comment by Yurlungur** [comment]
Oh one more thought. We might want to use `fprintf` to stderr and `KOKKOS_INLINE_FUNCTION` so it will work on device.

**Comment by brryan** [comment]
@Yurlungur yes the macros should call `exit`... oops. Also thanks for the work-on-device tip.

**Comment by JoshuaSBrown** [comment]
Nice Brryan!

**Comment by JoshuaSBrown** [comment]
I am wondering though if simply calling exit is the best solution, it might be better to throw so that we have the option to unwind the stack.

**Comment by brryan** [comment]
@JoshuaSBrown Thanks! I switched from `exit()` to `runtime_error`. If this isn't ideal (historically I've been more of a C99er) let me know.

@forrestglines I agree `DEBUG` versions of these calls that can be removed by the preprocessor are desirable, I've just added these. I'm only using a local `DEBUG_MODE_` flag in `error_checking.hpp` right now, though, and I think a unit test of this file, while good, won't work well without a global parthenon DEBUG macro. Have there been thoughts already on how/whether to put in such a flag?

**Comment by Yurlungur** [comment]
> @JoshuaSBrown Thanks! I switched from `exit()` to `runtime_error`. If this isn't ideal (historically I've been more of a C99er) let me know.

@JoshuaSBrown I understand the desire for a stack trace, but I would rather avoid throwing errors, or make it a switch set with a `#define`. `throw` will not work on the GPU and is also  carries a significant performance hit. I raised this issue in #81 

> @forrestglines I agree `DEBUG` versions of these calls that can be removed by the preprocessor are desirable, I've just added these. I'm only using a local `DEBUG_MODE_` flag in `error_checking.hpp` right now, though, and I think a unit test of this file, while good, won't work well without a global parthenon DEBUG macro. Have there been thoughts already on how/whether to put in such a flag?

We can perhaps implement a global `PARTHENON_DEBUG` macro. Alternatively we could use `#ifndef NDEBUG` so your code is disabled when `assert` is disabled. This would also have the advantage that `CMAKE_BUILD_TYPE=Debug` would enable your macro.

**Comment by JoshuaSBrown** [comment]
> `throw` will not work on the GPU and is also carries a significant performance hit.

I thought that might be an issue.

**Comment by brryan** [comment]
@JoshuaSBrown no worries, thanks for looking over this :)

@Yurlungur If you'd like me to add a `PARTHENON_DEBUG` I'm happy to do that in this PR.

**Comment by Yurlungur** [comment]
@brryan that feels like it belongs in a bigger discussion. For now, use `#ifndef NDEBUG`

**Comment by Yurlungur** [comment]
Looks good now, @brryan. One weird issue remaining: CI does not appear to have triggered. We might have to move your PR into a local branch so we get CI.

**Comment by brryan** [comment]
@Yurlungur Thanks Jonah, I did forget to run cpplint, I just did and fixed the errors. 

Yes I have not seen the CI run for this yet.

**Comment by Yurlungur** [comment]
@brryan Can you try making a branch in the lanl repository and seeing if that triggers CI? If the CI passes in the branch (and others approve), I think we can merge this in and then close the branch.

**Comment by Yurlungur** [comment]
I don't think code tests are required for this and the cpplint test says formatting is fine. Some I'm going to check off the other tasks.

**Comment by Yurlungur** [comment]
This looks like it's in good shape. Does anybody object to me merging once the checks pass?

**Comment by nmsriram** [comment]
Is it also worth crating a PARTHENON_WARN(level, message) macro so that we can adjust verbosity of runs?  This could also be turned off for non-debug runs.

**Comment by brryan** [comment]
@nmsriram I like the idea of `PARTHENON_WARN`/`PARTHENON_DEBUG_WARN` macros, just added this. Right now they only take `message` as an argument, I'm not sure what the best way to add verbosity levels would be (my preference now is to set this at compile time for performance). Others might have thoughts about this, maybe that could wait along with a `PARTHENON_DEBUG` flag for another PR?

**Comment by nmsriram** [comment]
@brryan thank you so much for adding those.  Agree on waiting for another PR.

**Comment by nmsriram** [comment]
@brryan, could you please identify why the checks are not passing and correct this.  If it is the case that GitHub cannot handle forks, please create a branch in the main repository and modify pull request accordingly.

**Comment by Yurlungur** [comment]
> @brryan, could you please identify why the checks are not passing and correct this. If it is the case that GitHub cannot handle forks, please create a branch in the main repository and modify pull request accordingly.

@brryan may be out for a while. I'll reach out to him and see if this can be done by someone else. I can shepherd it to completion if need be.

**Comment by Yurlungur** [comment]
@par-hermes format

**Comment by Yurlungur** [comment]
Please approve #153. It's the same but CI triggers and passes.

**Comment by Yurlungur** [comment]
Merged in as #153. Closing this one.

**Pull request closed without merging**